include/net/http_client.h

```diff
@@ -0,0 +1,240 @@
+ */
+typedef void (*http_response_cb_t)(struct http_response *rsp,
+				   enum http_final_call final_data,
+				   void *user_data);
+
+/**
+ * HTTP response from the server.
+ */
+struct http_response {
+	/** HTTP parser settings for the application usage */
+	const struct http_parser_settings *http_cb;
+
+	/** User provided HTTP response callback which is
+	 * called when a response is received to a sent HTTP
+	 * request.
+	 */
+	http_response_cb_t cb;
+
+	/** Where the body starts */
+	u8_t *body_start;
+
+	/** Where the response is stored, this is to be
+	 * provided by the user.
+	 */
+	u8_t *recv_buf;
+
+	/** Response buffer maximum length */
+	size_t recv_buf_len;
+
+	/** Length of the data in the result buf. If the value
+	 * is larger than recv_buf_len, then it means that
+	 * the data is truncated and could not be fully copied
+	 * into recv_buf. This can only happen if the user
+	 * did not set the response callback. If the callback
+	 * is set, then the HTTP client API will call response
+	 * callback many times so that all the data is
+	 * delivered to the user.
+	 */
+	size_t data_len;
+
+	/** HTTP Content-Length field value */
+	size_t content_length;
+
+	/** Content length parsed. This should be the same as
+	 * the content_length field if parsing was ok.
+	 */
+	size_t processed;
+
+	/* https://tools.ietf.org/html/rfc7230#section-3.1.2
+	 * The status-code element is a 3-digit integer code
+	 *
+	 * The reason-phrase element exists for the sole
+	 * purpose of providing a textual description
+	 * associated with the numeric status code. A client
+	 * SHOULD ignore the reason-phrase content.
+	 */
+	char http_status[HTTP_STATUS_STR_SIZE];
+
+	u8_t cl_present:1;
+	u8_t body_found:1;
+	u8_t message_complete:1;
+};
+
+/**
+ * HTTP client request. This contains all the data that is needed when doing
+ * a HTTP request.
+ */
+struct http_request {
+	/** HTTP parser context */
+	struct http_parser parser;
+
+	/** HTTP parser settings */
+	struct http_parser_settings parser_settings;
+
+	/** HTTP response specific data (filled by http_client_req() when
+	 * data is received)
+	 */
+	struct http_response response;
+
+	/** User data */
+	void *user_data;
+
+	/** HTTP socket */
+	int sock;
+
+	/** Request timeout */
+	s32_t timeout;
+
+	/* User should fill in following parameters */
+
+	/** The HTTP method: GET, HEAD, OPTIONS, POST, ... */
+	enum http_method method;
+
+	/** The URL for this request, for example: /index.html */
+	const char *url;
+
+	/** The HTTP protocol, for example "HTTP/1.1" */
+	const char *protocol;
+
+	/** The HTTP header fields (application specific)
+	 * The Content-Type may be specified here or in the next field.
+	 * Depending on your application, the Content-Type may vary, however
+	 * some header fields may remain constant through the application's
+	 * life cycle. This is a NULL terminated list of header fields.
+	 */
+	const char **header_fields;
```

Comment:
You create a HTTP request by setting things up in your application, then calling the API func. What do you mean by generating headers dynamically in this case, what kind of dynamical behaviour your application would need?

Reply:
> What do you mean by generating headers dynamically in this case, what kind of dynamical behaviour your application would need?

Well, suppose an app wants to send 2 headers with a payload of 2048 bytes for each (some auth-related material for example). Then, giving an option to send headers one by one would help. This could be achieved by providing flexible low-level API, allowing user to construct request piece by piece, e.g. request_start(), request_send_header(hdr, value) (can be called multiple times), request_send_payload() (can be called multiple times), request_end(). On top of that, "convenience" functions can be added for most common scenarios.


And it's just an idea I wanted to mention in review comments, I don't call it to be reworked like that so late in the review cycle. (Well, unless you think otherwise ;-) ).